### ISSUE TITLE
Problems with regex under non-neko

### ISSUE THREAD

**Issue by deanm** [haxiomic/haxe-glsl-parser#8]
I am using the parser from a macro, which unfortunately means it is running on some really bad fallback regex implementation.

The bad news is it took a while to debug, the good news is the changes are pretty minimal, but there were still a few things I couldn't get exactly right.

```
-       static var operatorRegex = ~/[&<=>|*?!+%(){}.~:,;\/\-\^\[\]]/;
+       static var operatorRegex = ~/[&<=>|*!?+%(){}.~:,;\/\^\[-]/;
```

So a few things here, the fallback regex things `*?` is a non-greedy qualifier, even though it is in a character class.  Secondly, no matter how many different ways I tried to escape it, I couldn't make `]` work in the character class.

My very unattractive solution was:

```
-               OPERATOR               => function() return operatorRegex.match(c(i))                      ? 1 : null,
+               OPERATOR               => function() return operatorRegex.match(c(i)) || c(i) == "]"       ? 1 : null,
```

Using a regex here is the wrong tool anyway, it is basically just a strchr(), so maybe it should just be an operatorString with a indexOf check instead.

Additionally `\s` isn't recognized, I tried to replace it with the proper character class definition, but there is no escapes for `\v` and `\f`.  Since there are not really ever used in practiced I just went with `[\t \n\r]`.  Additionally `\w` doesn't work, but can be replied by `[A-Za-z0-9_]`.

I didn't run all the tests, but this at least gets things running for me and parsing my files.

I'm more than happy to send out a patch, the only thing is the `\w` and `\s` stuff is ugly (and missing `\v` and `\f`), so perhaps someone knows of a better solution for that.

Thanks

**Comment by haxiomic** [comment]
Hey,

Great gob diagnosing this one

I totally agree regarding regex not being the right way to do this - just an array of chars + indexOf would be ideal.

Could even use the existing operatorMap:
```haxe
OPERATOR               => function(){
	for(op in operatorMap.keys()){
		if(op.charAt(0) == c(i)) return 1;
	}
	return null;
}
```
So we don't need to maintain two sets of character lists

I'm having trouble reproducing the regex issues however, when I try

```haxe
static var operatorRegex = ~/[&<=>|*?!+%(){}.~:,;\/\-\^\[\]]/;
static var whitespaceRegex = ~/\s/;
static var whitespaceNameMap = [
    ' ' => 'SPACE',
    '\t' => 'TAB'
];

static function main(){
    tryRegex('hello [world] 1 2 3\t4');
}

static macro function tryRegex(str:String){
    trace('Operator Regex on $str');

    operatorRegex.match(str);
    while(operatorRegex.match(str)){
        trace(operatorRegex.matched(0));
        str = operatorRegex.matchedRight();
    }

    trace('Whitespace Regex on $str');
    whitespaceRegex.match(str);
    while(whitespaceRegex.match(str)){
        trace(whitespaceNameMap.get(whitespaceRegex.matched(0)));
        str = whitespaceRegex.matchedRight();
    }

    return macro null;
}
```

At compile time it prints
```
: Operator Regex on hello [world] 1 2 3	4
: [
: ]
: Whitespace Regex on  1 2 3	4
: SPACE
: SPACE
: SPACE
: TAB
```
As expected

Are you testing on Windows?

The `\s` issue is surprising - this is something that should probably be fixed upstream in the regex implementation. It's weird it's not been fixed yet, since regex is often used in macros. Could you provide a minimal repro?

**Comment by deanm** [comment]
`\s` problem looked like:
```
> haxe-3.3.0-rc2/std/neko/_std/EReg.hx:33: characters 11-61 : Unsupported escaped char 's'
> <builtin>:1: character 0 : Called from
> haxe-3.3.0-rc2/std/neko/_std/EReg.hx:22: lines 22-204 : Called from
> haxe-glsl-parser/glsl/lex/Tokenizer.hx:424: characters 46-51 : Called from
> haxe-glsl-parser/glsl/lex/Tokenizer.hx:140: characters 10-21 : Called from
> haxe-glsl-parser/glsl/lex/Tokenizer.hx:177: characters 5-24 : Called from
> haxe-glsl-parser/glsl/lex/Tokenizer.hx:107: characters 23-38 : Called from
> haxe-glsl-parser/glsl/parse/Parser.hx:46: characters 15-40 : Called from
[ ... ]
> Aborted
```

**Comment by deanm** [comment]
Warning: I've looked at Haxe for the first time a few days ago, so I really don't know very much about it.  Anyway, I read this might have something to do with whether or not you have neko installed (I don't, I think).

https://github.com/HaxeFoundation/haxe/issues/4625
https://groups.google.com/forum/#!msg/haxelang/EkGQw5qCR-0/BAavEcEzDAAJ

Seems other projects have had problems, one example:

https://github.com/fponticelli/thx.core/issues/50

**Comment by haxiomic** [comment]
Yeah looks like that's the problem, haxe is pretty tied to neko (which is a VM that's also part of haxe) so it's worth installing

What are you aiming to build?

**Comment by deanm** [comment]
So actually I can't get it to use libneko, dtrace shows that it finds and opens the libneko.dylib, but then it doesn't seem to use it and I still get the same errors... It works fine with my patch so that's good enough for now, I don't really feel like trying to debug the neko stuff.

Thanks!
